test/Hamiltonians.jl

```diff
@@ -1723,3 +1724,58 @@ end
         @test eval(Meta.parse(repr(r))) == r
     end
 end
+
+@testset "HamiltonianProduct" begin
+    addr = BoseFS(2,0,0)
+
+    H = HubbardReal1D(addr)
+    start_at = DVec(addr => 10; style=IsStochasticWithThreshold(0.1))
+    P = H*H
+    problem = ProjectorMonteCarloProblem(P; start_at, last_step=10000, target_walkers=10000)
+    df = DataFrame(solve(problem))
+    energy = shift_estimator(df; skip=5000)
+    @test energy.mean ≈ eigvals(Matrix(P))[1] atol=5*energy.err
+
+    H1 = HubbardReal1D(addr;u=1.0im)
+    H2 = ExtendedHubbardReal1D(addr)
+    @test LOStructure(H2*H2) == IsHermitian()
+    P = H1*H2
+    @test LOStructure(P) == AdjointKnown()
+    c = operator_column(P, addr)
+
+    for _ in 1:20
+        a, p, v = random_offdiagonal(c)
+        @test (a => v) in offdiagonals(c)
+    end
+
+    ods = [DVec(od) for od in offdiagonals(c)]
+    ods_product = DVec(addr => diagonal_element(c)) + sum(ods)
```

Comment:
Would this be equivalent? 
```suggestion
    ods_product = DVec(offdiagonals(c))
    ods_product[addr] += diagonal_element(c)
```
I find it easier to parse.

Reply:
Ok, I've changed it to be more clear. `DVec(offdiagonals(c)` doesn't work though because there are repeats (different paths to the same address), and only one gets picked.

Reply:
Hmm, you are right.  `DVec(offdiagonals(c)` doesn't work because it first converts the arguments into a `Dict`. I'm thinking whether this is what we want or whether we should treat this behaviour as a bug.